nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -480,13 +483,116 @@ def forward(
+        context_layer = context_layer.view(*new_context_layer_shape)
+        return context_layer
+
+
+class FlashAttention(MegatronModule):
+
+    def __init__(
+        self,
+        layer_number,
+        num_attention_heads,
+        hidden_size,
+        attention_type=AttnType.self_attn,
+        attn_mask_type=AttnMaskType.padding,
+        precision=16,
+        kv_channels=None,
+        attention_dropout=0.1,
+        sequence_parallel=False,
+
+    ):
+
+        super(FlashAttention, self).__init__()
+
+        if precision == 32:
+            raise ValueError('FlashAttention does not support fp32.')
+
+        self.precision = precision
+        self.fp16 = precision == 16
+        self.bf16 = precision == 'bf16'
+
+        if self.fp16:
+            self.dtype = torch.float16
+        else:
+            self.dtype = torch.bfloat16
+
+        self.flash_attention_fn = flash_attn_unpadded_qkvpacked_func
+    
+        self.layer_number = max(1, layer_number)
+        self.attention_type = attention_type
+        self.attn_mask_type = attn_mask_type
+        self.sequence_parallel = sequence_parallel
+
+        if kv_channels is None:
+            assert (
+                hidden_size % num_attention_heads == 0
+            ), 'hidden_size must be divisible by num_attention_heads if kv_channels is None'
+            kv_channels = hidden_size // num_attention_heads
+
+        projection_size = kv_channels * num_attention_heads
+
+        # Per attention head and per partition values.
+        world_size = parallel_state.get_tensor_model_parallel_world_size()
+        self.hidden_size_per_partition = safe_divide(projection_size, world_size)
+        self.hidden_size_per_attention_head = safe_divide(projection_size, num_attention_heads)
+        self.num_attention_heads_per_partition = safe_divide(num_attention_heads, world_size)
+        self.num_attention_heads_partition_offset = (
+            self.num_attention_heads_per_partition * parallel_state.get_tensor_model_parallel_rank()
+        )
+
+        self.attention_dropout = attention_dropout
+
+
+
+    def forward(self, query_layer, key_layer, value_layer, attention_mask, **kwargs):
+        # [b, np, sq, sk]
+        output_size = (
+            query_layer.size(1),
+            query_layer.size(2),
+            query_layer.size(0),
+            key_layer.size(0),
+        )
+        causal = attention_mask is not None
+        # [s, b, np, hn] -> [b, s, np, hn] -> [b * s, 1, np, hn]
+        query_layer = query_layer.transpose(0, 1).reshape(output_size[0] * output_size[2], 1, output_size[1], -1)
+        key_layer = key_layer.transpose(0, 1).reshape(output_size[0] * output_size[3], 1, output_size[1], -1)
+        value_layer = value_layer.transpose(0, 1).reshape(output_size[0] * output_size[3], 1, output_size[1], -1)
+
+        # Combined q/k/v into [b * s, 3, np, hn].
+        qkv = torch.concat([query_layer, key_layer, value_layer], dim=1)
+        prev_dtype = None
```

Comment:
@Quentin-Anthony I'd like to draw your attention to this block. FlashAttention does not support fp32, but all of the NeMo Transformer code fp16 or bf16 operations are handled by AMP autocasting. Because FlashAttention uses custom CUDA kernels, AMP doesn't work and I needed to do this check and cast in case the linear layers were initialized to fp32. Does this look ok to you?